README.rst
===================
kotti_image_gallery
===================

This is an extension to the Kotti CMS that allows you to add galleries with images to your Kotti site.

It uses `Bootstrap Carousel`_ for the gallery view.

`Find out more about Kotti`_


Setup
=====

To activate the ``kotti_image_gallery`` add-on in your Kotti site, you need to add an entry to the ``kotti.configurators`` setting in your Paste Deploy config.
If you don't have a ``kotti.configurators`` option, add one.
The line in your ``[app:main]`` section could then look like this::

  kotti.configurators = kotti_image_gallery.kotti_configure

With this, you'll be able to add gallery and image items in your site.


Image URLs
==========

``kotti_image_gallery`` provides on-the-fly image scaling by utilizing `plone.scale`_ (thanks to Tom Lazar for the hint).

Images (including arbitrary scales) can be referenced by this URL schema: ``/path/to/image_content_object/image[[/<image_scale>]/download]`` where ``<image_scale>`` can be:

 - a predefined image scale (see below)
 - a string of the form ``<max_width>x<max_height>`` or
 - a URL path segment of the form ``<max_width>/<max_height>``

If the last URL path segment is ``download``, the image will be served with ``Content-disposition: attachment`` otherwise it will be served with ``Content-disposition: inline``.

Predefined image scale sizes
----------------------------

You may define image scale sizes in your ``.ini`` file by setting values for ``kotti_image_gallery.scale_<scale_name>`` to values of the form ``<max_width>x<max_height>`` (e.g. ``kotti_image_gallery.scale_thumb = 160x120`` with the resulting scale name ``thumb``).

``thumb`` (160x120) and ``carousel`` (560x420) are always defined (because they are used in the default templates), but their values can be overwritten by setting ``kotti_image_gallery.scale_thumb`` and/or ``kotti_image_gallery.scale_carousel`` to different values in your .ini file.


Work in progress
================

``kotti_image_gallery`` is considered alpha software, not suitable for use in production environments.
The current state of the project is in no way feature complete nor API stable.
If you really want to use it in your project(s), make sure to pin the exact version in your requirements.
Not doing so will likely break your project when future releases become available.


Development
===========

Contributions to ``kotti_image_gallery`` are highly welcome.
Just clone its `Github repository`_ and submit your contributions as pull requests.

Note that all development is done on the ``develop`` branch and ``master`` is reserved for "production-ready state".
Therefore make sure to always base your work on the current state of the ``develop`` branch.

This follows the highly recommended `A successful Git branching model`_ pattern, which is implemented by the excellent `gitflow`_ git extension.

Testing
-------

``kotti_image_gallery`` has 100% test coverage.
Please make sure that you add tests for new features and that all tests pass before submitting pull requests.
Running the test suite is as easy as running ``py.test`` from the source directory.


.. _Bootstrap Carousel: http://twitter.github.com/bootstrap/javascript.html#carousel
.. _Find out more about Kotti: http://pypi.python.org/pypi/Kotti
.. _`plone.scale`: http://pypi.python.org/pypi/plone.scale/1.2.2
.. _Github repository: https://github.com/disko/kotti_image_gallery
.. _gitflow: https://github.com/nvie/gitflow
.. _A successful Git branching model: http://nvie.com/posts/a-successful-git-branching-model/


CHANGES.txt
CHANGES
=======

0.3.2
-----

- added missing dependency (PIL)

0.3.1
-----

- fix packaging (thanks to Marco Scheidhuber for reporting)

0.3
---

- don't duplicate code/functionality that already exists in Kotti
- 100% test coverage

0.2
---

- Image inherits from File now.
- Replaced repoze.bitblt based image scaling with much more powerful plone.scale based version.

0.1
---

- Initial release.


kotti_image_gallery/tests/test_settings.py
# -*- coding: utf-8 -*-

from kotti.testing import UnitTestBase

class SettingsTests(UnitTestBase):

    def test_01_defaults(self):

        settings = {'kotti.includes': '',
                    'kotti.available_types': '', }

        import kotti_image_gallery

        kotti_image_gallery.kotti_configure(settings)

        self.assertTrue(settings['kotti.includes'].find('kotti_image_gallery.views') >= 0)
        self.assertTrue(settings['kotti.available_types'].find('kotti_image_gallery.resources.Gallery') >= 0)
        self.assertTrue(settings['kotti.available_types'].find('kotti_image_gallery.resources.Image') >= 0)

        self.assertEquals(kotti_image_gallery.image_scales['thumb'], [160, 120])
        self.assertEquals(kotti_image_gallery.image_scales['carousel'], [560, 420])

    def test_02_override_defaults(self):

        settings = {'kotti.includes': '',
                    'kotti.available_types': '',
                    'kotti_image_gallery.scale_thumb' : '100x200',
                    'kotti_image_gallery.scale_carousel' : '300x400', }

        import kotti_image_gallery

        kotti_image_gallery.kotti_configure(settings)

        self.assertEquals(kotti_image_gallery.image_scales['thumb'], [100, 200])
        self.assertEquals(kotti_image_gallery.image_scales['carousel'], [300, 400])

    def test_03_additional_scale(self):

        settings = {'kotti.includes': '',
                    'kotti.available_types': '',
                    'kotti_image_gallery.scale_huge' : '1000x2000', }

        import kotti_image_gallery

        kotti_image_gallery.kotti_configure(settings)

        self.assertEquals(kotti_image_gallery.image_scales['huge'], [1000, 2000])

    def test_04_ivalid_scale(self):

        settings = {'kotti.includes': '',
                    'kotti.available_types': '',
                    'kotti_image_gallery.scale_invalid' : '100,200', }

        import kotti_image_gallery

        kotti_image_gallery.kotti_configure(settings)

        self.assertTrue('invalid' not in kotti_image_gallery.image_scales)


kotti_image_gallery/resources.py
# -*- coding: utf-8 -*-

from kotti.resources import Content, File
from kotti.util import _
from sqlalchemy import Column, ForeignKey, Integer


class Gallery(Content):

    id = Column(Integer, ForeignKey('contents.id'), primary_key=True)

    type_info = Content.type_info.copy(name=u'Gallery',
                                       title=_(u'Gallery'),
                                       add_view=u'add_gallery',
                                       addable_to=[u'Document'], )


class Image(File):

    id = Column(Integer(), ForeignKey('files.id'), primary_key=True)

    type_info = File.type_info.copy(name=u'Image',
                                    title=_(u'Image'),
                                    add_view=u'add_image',
                                    addable_to=[u'Gallery', u'Document', u'Video'], )


kotti_image_gallery/views.py
# -*- coding: utf-8 -*-

import PIL
from kotti import DBSession
from kotti.util import _
from kotti.views.edit import ContentSchema, make_generic_add, make_generic_edit
from kotti.views.file import AddFileFormView, EditFileFormView
from kotti_image_gallery import image_scales
from kotti_image_gallery.resources import Gallery, Image
from plone.scale.scale import scaleImage
from pyramid.response import Response
from pyramid.view import view_config

PIL.ImageFile.MAXBLOCK = 33554432

class GallerySchema(ContentSchema):
    pass


class BaseView(object):

    def __init__(self, context, request):

        self.context = context
        self.request = request


class GalleryView(BaseView):

    @view_config(context=Gallery,
                 name='view',
                 permission='view',
                 renderer='templates/gallery-view.pt')
    def view(self):

        session = DBSession()
        query = session.query(Image).filter(Image.parent_id==self.context.id).order_by(Image.position)
        images = query.all()

        return {"images": images}


class EditImageFormView(EditFileFormView):

    pass


class AddImageFormView(AddFileFormView):

    item_type = _(u"Image")

    def add(self, **appstruct):

        buf = appstruct['file']['fp'].read()

        return Image(title=appstruct['title'],
                     description=appstruct['description'],
                     data=buf,
                     filename=appstruct['file']['filename'],
                     mimetype=appstruct['file']['mimetype'],
                     size=len(buf), )


class ImageView(BaseView):

    @view_config(context=Image,
                 name='view',
                 permission='view',
                 renderer='templates/image-view.pt')
    def view(self):
        return {}

    @view_config(context=Image,
                 name="image",
                 permission='view')
    def image(self):
        """return the image in a specific scale, either inline (default) or as attachment"""

        subpath = list(self.request.subpath)

        if (len(subpath) > 0) and (subpath[-1] == "download"):
            disposition = "attachment"
            subpath.pop()
        else:
            disposition = "inline"

        if len(subpath) == 1:
            scale = subpath[0]
            if scale in image_scales:
                # /path/to/image/scale/thumb
                width, height = image_scales[scale]
            else:
                # /path/to/image/scale/160x120
                try:
                    width, height = [int(v) for v in scale.split("x")]
                except ValueError:
                    width, height = (None, None)

        elif len(subpath) == 2:
            # /path/to/image/scale/160/120
            try:
                width, height = [int(v) for v in subpath]
            except ValueError:
                width, height = (None, None)

        else:
            # don't scale at all
            width, height = (None, None)

        if width and height:
            image, format, size = scaleImage(self.context.data,
                                             width=width,
                                             height=height,
                                             direction="thumb")
        else:
            image = self.context.data

        res = Response(
            headerlist=[('Content-Disposition', '%s;filename="%s"' % (disposition,
                                                                      self.context.filename.encode('ascii', 'ignore'))),
                        ('Content-Length', str(len(image))),
                        ('Content-Type', str(self.context.mimetype)), ],
            app_iter=image)

        return res

def includeme(config):

    config.add_static_view('static-kotti_image_gallery', 'kotti_image_gallery:static')
    config.scan("kotti_image_gallery")
    config.add_view(AddImageFormView,
                    name=Image.type_info.add_view,
                    permission='add',
                    renderer='kotti:templates/edit/node.pt',)
    config.add_view(EditImageFormView,
                    context=Image,
                    name='edit',
                    permission='edit',
                    renderer='kotti:templates/edit/node.pt', )

    config.add_view(make_generic_edit(GallerySchema()),
                    context=Gallery,
                    name='edit',
                    permission='edit',
                    renderer='kotti:templates/edit/node.pt', )

    config.add_view(make_generic_add(GallerySchema(), Gallery),
                    name=Gallery.type_info.add_view,
                    permission='add',
                    renderer='kotti:templates/edit/node.pt', )
